Ghidra/Processors/x86/data/languages/lockable.sinc

```diff
@@ -0,0 +1,326 @@
+    build lockx;
+@ifdef IA64
+    local ptr = Mem + (sext(Reg32) s>>3);
+@else
+    local ptr = Mem + (Reg32 s>>3);
+@endif
+    local bit = Reg32 & 7;
+    local val = (*:1 ptr >> bit) & 1;
+    *:1 ptr = *:1 ptr | (1<<bit);
+    CF = (val != 0);
+    build unlock;
+}
+@ifdef IA64
+:BTS^lockx^unlock Mem,Reg64	is $(LONGMODE_ON) & vexMode=0 & lockx & unlock & opsize=2 & byte=0xf; byte=0xab; Mem & Reg64 ...		{ build lockx; local ptr = Mem + (Reg64 s>>3); local bit = Reg64 & 7; local val = (*:1 ptr >> bit) & 1; *:1 ptr = *:1 ptr | (1<<bit); CF = (val != 0); build unlock; }
+@endif
+:BTS^lockx^unlock m16,imm8     is vexMode=0 & lockx & unlock & opsize=0 & byte=0xf; byte=0xba; m16 & reg_opcode=5 ...; imm8   { build lockx; local bit=imm8&0xf; local val=(m16>>bit)&1; m16=m16 | (1<<bit); CF=(val!=0); build unlock; }
+:BTS^lockx^unlock m32,imm8     is vexMode=0 & lockx & unlock & opsize=1 & byte=0xf; byte=0xba; m32 & reg_opcode=5 ...; imm8   { build lockx; local bit=imm8&0x1f; local val=(m32>>bit)&1; CF=(val!=0); m32=m32 | (1<<bit); build unlock; }
+@ifdef IA64
+:BTS^lockx^unlock m64,imm8     is $(LONGMODE_ON) & vexMode=0 & lockx & unlock & opsize=2 & byte=0xf; byte=0xba; m64 & reg_opcode=5 ...; imm8   { build lockx; local bit=imm8&0x3f; local val=(m64>>bit)&1; m64=m64 | (1<<bit); CF=(val!=0); build unlock; }
+@endif
+
+:CMPXCHG^lockx^unlock m8,Reg8  is vexMode=0 & lockx & unlock & byte=0xf; byte=0xa6; m8 & Reg8 ...           { build lockx; build unlock; }
+:CMPXCHG^lockx^unlock m8,Reg8  is vexMode=0 & lockx & unlock & byte=0xf; byte=0xb0; m8 & Reg8 ...           { build lockx; subflags(AL,m8); local tmp=AL-m8; resultflags(tmp);
+                                          local diff = m8^Reg8; m8 = m8 ^ (ZF*diff);
+                                          diff = AL ^ m8; AL = AL ^ ((ZF==0)*diff);
+					  build unlock; }
+:CMPXCHG^lockx^unlock m16,Reg16    is vexMode=0 & lockx & unlock & opsize=0 & byte=0xf; byte=0xb1; m16 & Reg16 ...  { build lockx; subflags(AX,m16); local tmp=AX-m16; resultflags(tmp);
+                                          local diff = m16^Reg16; m16 = m16 ^ (zext(ZF) * diff);
+                                          diff = AX ^ m16; AX = AX ^ (zext(ZF==0) * diff);
+					  build unlock; }
+:CMPXCHG^lockx^unlock m32,Reg32    is vexMode=0 & lockx & unlock & opsize=1 & byte=0xf; byte=0xb1; m32 & Reg32 ... & check_EAX_dest ...
+{
+    build lockx;
+	#this instruction writes to either EAX or m32
+	#in 64-bit mode, a 32-bit register that is written to
+	#(and only the register that is written to)
+	#must be zero-extended to 64 bits
+	subflags(EAX,m32);
+	local tmp=EAX-m32;
+	resultflags(tmp);
+    if (ZF==1) goto <equal>;
+    EAX = m32;
+    build check_EAX_dest;
+    #TODO Errors on duplicate BUILD statements
+    #build unlock;
+    goto inst_next;
+<equal>
+    m32 = Reg32;
+    #TODO Errors on duplicate BUILD statements
+    #build unlock;
+}
```

Comment:
I'm not sure how to resolve the issue of duplicate BUILD statements here

Reply:
you could add one more label: (same number of gotos)

```
#build unlock; # remove this
goto <inst_end>; # replace goto inst_next;
<equal>
m32 = Reg32;
<inst_end>
build unlock;
```